Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm

```diff
@@ -779,23 +833,22 @@ - (void)drawInContext:(CGContextRef)cgContext
     auto startTime = MonotonicTime::now();
 
     do {
-        switch (m_webPageProxyProcessState.commitLayerTreeMessageState) {
-        case CommitLayerTreePending:
-        case MissedCommit:
-        case Idle:
-            if (!m_webPageProxyProcessState.receivedCommitLayerTreePendingConfirmation) {
-                if (connection->waitForAndDispatchImmediately<Messages::RemoteLayerTreeDrawingAreaProxy::NotifyPendingCommitLayerTree>(identifier(), activityStateUpdateTimeout - (MonotonicTime::now() - startTime), IPC::WaitForOption::InterruptWaitingIfSyncMessageArrives) != IPC::Error::NoError)
-                    return;
+        IPC::Error error = IPC::Error::NoError;
+        WTF::switchOn(m_webPageProxyProcessState.commitLayerTreeMessageState,
+            [&](const CommitLayerTreePending& commitLayerTreePending) {
+                if (commitLayerTreePending.requestedNotifyPendingCommitLayerTree)
+                    error = connection->waitForAndDispatchImmediately<Messages::RemoteLayerTreeDrawingAreaProxy::NotifyPendingCommitLayerTree>(identifier(), activityStateUpdateTimeout - (MonotonicTime::now() - startTime), IPC::WaitForOption::InterruptWaitingIfSyncMessageArrives);
+                else
+                    error = connection->waitForAndDispatchImmediately<Messages::RemoteLayerTreeDrawingAreaProxy::CommitLayerTree>(identifier(), activityStateUpdateTimeout - (MonotonicTime::now() - startTime), IPC::WaitForOption::InterruptWaitingIfSyncMessageArrives);
+            }, [&](const NeedsDisplayDidRefresh&) {
+                didRefreshDisplay(connection.ptr());
```

Comment:
Totally optional food for thought:
This variant is not like the others... Would it be worth, and would it even make sense, to have `didRefreshDisplay` report the `IPC::Error` from its `connection.send()`?

If yes, this whole `SwitchOn` expression could consistently return an error from all branches, so you wouldn't need the separate init line above. But once again I'm just looking at the repetitive pattern, it may not make sense to force the `NeedsDisplayDidRefresh` case to do the same as the others.